lib/net/ssh/transport/aead_aes_gcm.rb

```diff
@@ -0,0 +1,60 @@
+require 'openssl'
+require 'delegate'
+
+module Net::SSH::Transport
+  # :nodoc:
+  # Ruby implementation of AEAD Aes-GCM Mode
+  # for Block Ciphers. See RFC5647 for details.
+  # The main purpose is the implementation of the GCM iv increment
+  module AEADAESGCM
+    def self.extended(orig)
+      orig.instance_eval do
+        @iv = { fixed: nil, invocation_ctr: nil }
+        orig.padding = 0
+
+        singleton_class.send(:alias_method, :_update, :update)
+        singleton_class.send(:private, :_update)
+        singleton_class.send(:undef_method, :update)
+
+        def self.block_size
+          16
+        end
+
+        def iv_len
+          12
+        end
+
+        def iv=(iv_s)
+          if @iv[:fixed].nil?
+            @iv[:fixed] = iv_s[0...4]
+            @iv[:invocation_ctr] = iv_s[4..12]
+          end
+          super(iv_s)
+        end
+
+        def incr_iv
+          return if @iv[:fixed].nil?
+
+          @iv[:invocation_ctr] = [(@iv[:invocation_ctr].unpack1('B*').to_i(2) + 1)].pack('Q>*')
+          self.iv = "#{@iv[:fixed]}#{@iv[:invocation_ctr]}"
+        end
+
+        def padding=(pad)
+          # DO NOTHING (always 0)
```

Comment:
Minor nit: this comment could be nicer to readers and explain that the padding _content_ is 0 and this does not refer to padding length.